awscli/customizations/codeartifact/login.py

```diff
@@ -162,59 +165,43 @@ def login(self, dry_run=False):
         self._write_success_message('nuget')
 
     def _get_source_to_url_dict(self):
-        # The response from listing sources takes the following form:
-        #
-        # Registered Sources:
-        #   1.  Source Name 1 [Enabled]
-        #       https://source1.com/index.json
-        #   2.  Source Name 2 [Disabled]
-        #       https://source2.com/index.json
-        # ...
-        #   100. Source Name 100
-        #       https://source100.com/index.json
-
-       # Or it can be (blank line after Registered Sources:)
-
-       # Registered Sources:
-
-       #   1.  Source Name 1 [Enabled]
-       #       https://source1.com/index.json
-       #   2.  Source Name 2 [Disabled]
-       #       https://source2.com/index.json
-       # ...
-       #   100. Source Name 100
-       #       https://source100.com/index.json
+        """
+        Parses the output of the nuget sources list command.
 
+        A dict is created where the keys are the source names
+        and the values the corresponding URL.
+
+        The output of the command can contain header and footer information
+        around the 'Registered Sources' section, which is ignored.
+
+        Example output that is parsed:
+
+        Registered Sources:
+
+        1. Source Name 1 [Enabled]
+           https://source1.com/index.json
+        2. Source Name 2 [Disabled]
+           https://source2.com/index.json
+        100. Source Name 100 [Activé]
+             https://source100.com/index.json
+        """
         response = self.subprocess_utils.check_output(
             self._get_list_command(),
             stderr=self.subprocess_utils.PIPE
         )
 
-        lines = response.decode("utf-8").splitlines()
+        lines = response.decode(os.device_encoding(1) or "utf-8").splitlines()
```

Comment:
Is there a particular reason we are not using [`locale.getpreferredencoding()`](https://docs.python.org/3/library/locale.html#locale.getpreferredencoding) here? For determining encoding in the past, we've generally used this method to determine encoding preferences, but not sure if that function matches with how the dotnet CLI determines the encoding of its output. In general, we should be trying to match the encoding that the dotnet uses for output.

Reply:
The dotnet CLI returns output in the language specified by the Windows display language , not the "Language for non-Unicode programs"(system locale)). So that is why `locale.getpreferredencoding()` would not help here.

Reply:
Reading over the documentation, I agree `locale.getpreferredencoding()` is not what we want as it respects Python environment variables (which the dotnet CLI would not respect).

> The dotnet CLI returns output in the language specified by the Windows display language

But when it encodes the character values (based on the language selected) to binary/bits, how does it determine which encoding to use? Based on the updated logic, it would imply that the dotnet CLI encodes the characters to binary based on the encoding of the stdout file descriptor. I just want confirmation that is how the dotnet CLI determines which encoding to use since when the CLI subprocesses out to the dotnet CLI, the dotnet CLI's returned output will be in binary and the AWS CLI needs to decode the binary using the same encoding as the dotnet CLI to avoid any decoding issues.

Reply:
For Windows, as dotnet CLI returns output in the language specified by the Windows display language , not the "Language for non-Unicode programs", Below is the output 
```>>> import os
>>> import subprocess
>>> os.device_encoding(1)
'cp437'
>>> response = subprocess.check_output(['dotnet', 'nuget', 'list', 'source', '--format', 'detailed'])
>>> response
b'Sources inscrites\xff:\r\n  1.  nuget.org [Activ\x82]\r\n      https://api.nuget.org/v3/index.json\r\n'
>>> response.decode("utf-8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 17: invalid start byte
>>> response.decode(os.device_encoding(1) or "utf-8")
'Sources inscrites\xa0:\r\n  1.  nuget.org [Activé]\r\n      https://api.nuget.org/v3/index.json\r\n'
```
Also when i changed the system locale to `en`, i get the same output as above but if i remove to get console coding(os.device_encoding(1)), decoding issue pops. The shell (cmd, pwsh, bash, etc) isn’t relevant to the encoding the child process expects, but the console (conhost) might be if the current process is attached to the same console as the child process.Each process that attaches to a console shares the console’s active input and output codepages – i.e. the console doesn’t use a separate set of active codepages for each attached process. In Python, the console’s active output codepage is os.device_encoding(1) and encoding dotnet cli command uses are likely to be the same. 

Plus there is an issue open with dotnet to "Use UTF-8 Encoding by Default On Windows On Non English Languages & Change Culture For All Based on OS Main Language" which might solve the problem windows users are getting with console output. I also see in that above mentioned github issue code change, they are using `chcp` command to set which changes the code page which confirms this.

Reply:
Sounds good. Thanks for the confirmation.